// @microsoft/rush-lib/lib/cli/actions/CheckAction.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const BaseRushAction_1 = require("./BaseRushAction");
const VersionMismatchFinder_1 = require("../../api/VersionMismatchFinder");
class CheckAction extends BaseRushAction_1.BaseRushAction {
    constructor(parser) {
        super({
            actionName: 'check',
            summary: 'Checks each project\'s package.json files and ensures that all dependencies are of the same ' +
                'version throughout the repository.',
            documentation: 'Checks each project\'s package.json files and ensures that all dependencies are of the ' +
                'same version throughout the repository.',
            safeForSimultaneousRushProcesses: true,
            parser
        });
    }
    onDefineParameters() {
        // abstract
    }
    run() {
        VersionMismatchFinder_1.VersionMismatchFinder.rushCheck(this.rushConfiguration);
        return Promise.resolve();
    }
}
exports.CheckAction = CheckAction;
