// @microsoft/sp-dialog/lib/AlertDialog.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { DialogContent, DialogFooter } from 'office-ui-fabric-react/lib/components/Dialog';
import * as React from 'react';
import * as ReactDom from 'react-dom';
import { PrimaryButton } from '@microsoft/office-ui-fabric-react-bundle';
import BaseDialog from './BaseDialog';
/**
 * The React wrapper to use Fabric React components to render the alert dialog
 */
var AlertDialogReact = /** @class */ (function (_super) {
    __extends(AlertDialogReact, _super);
    function AlertDialogReact() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AlertDialogReact.prototype.render = function () {
        // Return the content of the dialog
        return React.createElement(DialogContent, { title: 'Alert', subText: this.props.message },
            React.createElement(DialogFooter, null,
                React.createElement(PrimaryButton, { onClick: this.props.close, text: 'OK' })));
    };
    return AlertDialogReact;
}(React.Component));
/**
 * Internal implementation of a dialog to mimic browser's alert() behavior.
 * Third-party can use this by calling the static method Dialog.alert()
 *
 * @internal
 */
var AlertDialog = /** @class */ (function (_super) {
    __extends(AlertDialog, _super);
    function AlertDialog(message) {
        var _this = _super.call(this, { isBlocking: false }) || this;
        _this._message = message;
        return _this;
    }
    AlertDialog.prototype.render = function () {
        ReactDom.render(React.createElement(AlertDialogReact, {
            message: this._message,
            close: this.close
        }), this.domElement);
    };
    AlertDialog.prototype.onAfterClose = function () {
        ReactDom.unmountComponentAtNode(this.domElement);
    };
    return AlertDialog;
}(BaseDialog));
export default AlertDialog;
